packages/ref-doc/src/utils/type-signature.ts

```diff
@@ -0,0 +1,143 @@
+    case "Projection":
+      return "(projection)";
+    case "Object":
+      return "(object)";
+    default:
+      const _assertNever: never = type;
+      compilerAssert(false, "Unexpected type kind");
+  }
+}
+
+function isReflectionType(type: Type): type is Model & { name: "" } {
+  return (
+    type.kind === "Model" &&
+    type.namespace?.name === "Reflection" &&
+    type.namespace?.namespace?.name === "Cadl"
+  );
+}
+
+function getDecoratorSignature(type: Decorator) {
+  const ns = getQualifier(type.namespace);
+  const name = type.name.slice(1);
+  const parameters = [type.target, ...type.parameters].map((x) => getFunctionParameterSignature(x));
+  return `dec ${ns}${name}(${parameters.join(", ")})`;
+}
+
+function getFunctionSignature(type: FunctionType) {
+  const ns = getQualifier(type.namespace);
+  const parameters = type.parameters.map((x) => getFunctionParameterSignature(x));
+  return `fn ${ns}${type.name}(${parameters.join(", ")}): ${getTypeName(type.returnType)}`;
+}
+
+function getInterfaceSignature(type: Interface) {
+  const ns = getQualifier(type.namespace);
+
+  const templateParams = type.node.templateParameters
+    ? getTemplateParameters(type.node.templateParameters)
+    : "";
+  return `interface ${ns}${type.name}${templateParams}`;
+}
+
+function getTemplateParameters(templateParameters: readonly TemplateParameterDeclarationNode[]) {
+  const params = templateParameters.map((x) => `${x.id.sv}`);
+  return `<${params.join(", ")}>`;
+}
+function getOperationSignature(type: Operation) {
+  const qualifier = getQualifier(type.interface ?? type.namespace);
+  const parameters = [...type.parameters.properties.values()].map(getModelPropertySignature);
+  return `op ${qualifier}${type.name}(${parameters.join(", ")}): ${getTypeName(type.returnType)}`;
+}
+
+function getFunctionParameterSignature(parameter: FunctionParameter) {
+  const rest = parameter.rest ? "..." : "";
+  const optional = parameter.optional ? "?" : "";
+  return `${rest}${parameter.name}${optional}: ${getTypeName(parameter.type)}`;
+}
+
+function getModelPropertySignature(property: ModelProperty) {
+  const ns = getQualifier(property.model);
+  return `${ns}${property.name}: ${getTypeName(property.type)}`;
+}
+
+function getUnionVariantSignature(variant: UnionVariant) {
+  if (typeof variant.name !== "string") {
+    return getTypeName(variant.type);
+  }
+  const ns = getQualifier(variant.union);
+  return `${ns}${variant.name}: ${getTypeName(variant.type)}`;
+}
+
+function getEnumMemberSignature(member: EnumMember) {
+  const ns = getQualifier(member.enum);
+  const value = typeof member.value === "string" ? `"${member.value}"` : member.value;
+  return value === undefined ? `${ns}${member.name}` : `${ns}${member.name}: ${value}`;
+}
+
+export function getQualifier(parent: Type | undefined) {
+  if (!parent) {
+    return "";
+  }
+
+  const parentName = getTypeName(parent);
```

Comment:
I was just noticing in the server copy of this where operation (not decorator or function) signatures show `(anonymous model).paramName`. I was going to fix it. Not for right now, but there is probably some opportunity to consolidate signature printing with some callbacks or options, not sure.

It looks like maybe we don't show interface operations yet in the ref docs? I tried to see if this was impacting this too but in the REST docs for example, I see the interfaces but not the operations in them.

Reply:
yeah for now I didn't show it, in the same way as I don't show model properties, enum members, etc.. I think that can be a follow up as might need a bit of dicussion on how we want to show it